client/Lykke.Service.Assets.Client/ReadModels/InMemoryAssetPairsReadModelRepository.cs

```diff
@@ -0,0 +1,52 @@
+﻿using Autofac;
+using Lykke.Service.Assets.Client.Models.v3;
+using Microsoft.Extensions.Caching.Memory;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Lykke.Service.Assets.Client.ReadModels
+{
+    class InMemoryAssetPairsReadModelRepository : IAssetPairsReadModelRepository, IStartable
+    {
+        public const string AllKey = "AllAssetPairs";
+
+        private readonly IAssetsService _assetsService;
+        private readonly IMemoryCache _cache;
+
+        public InMemoryAssetPairsReadModelRepository(IAssetsService assetsService, IMemoryCache cache)
+        {
+            _assetsService = assetsService;
+            _cache = cache;
+        }
+
+        public AssetPair TryGet(string id)
+        {
+            try
+            {
+                if (!_cache.TryGetValue(id, out AssetPair value))
+                    return null;
+                return value;
+            }
+            catch (System.InvalidCastException)
+            {
+                return null;
+            }
+        }
+
+        public IReadOnlyCollection<AssetPair> GetAll()
+        {
+            var ids = _cache.Get<List<string>>(AllKey);
+            return ids.Select(x => _cache.Get<AssetPair>(x)).ToArray();
```

Comment:
Is this operation thread safe?

Reply:
Well, it is written that MS-provided MemoryCache is entirely thread safe.

Reply:
Btw, it's not thread safe, because in methods, where you updating the cache, you insert/update AllKeys first and then insert asset/asset pair. Thus, due to race condition, ID could be pressented in the AllKey list, but asset/asset pair could be still not added to the cache.